File: components/RequestDispatchingRules/RequestDispatchingSimple.py
import json
import random


class RequestDispatchingSimple:
    def __init__(self):
        self.network_devices = []

    @classmethod
    def from_json(cls, json_data):
        """
        从 JSON 字符串或 JSON 文件内容初始化实例。
        参数：
            - json_data：JSON 格式的字符串或字典对象。
        """
        if isinstance(json_data, str):
            # 将 JSON 字符串转换为列表
            data = json.loads(json_data)
        else:
            raise ValueError("Invalid JSON data. Must be a JSON string.")

        return cls(
            network_devices=data
        )

    def to_json(self):
        """
        将实例转换为 JSON 格式的字符串。
        """

        # 转换为 JSON 字符串并返回
        return json.dumps(self.network_devices, indent=4)


    def find_device_id(self, this_device_id, request_can_dispatch_list, child_device_ids, parent_device_ids, same_level_device_ids):
        if not request_can_dispatch_list:
            print("Error - Service Discovery Information Missing or there is no service instance available.")
            return -1

        # 1. check self
        key_to_check = str(this_device_id)
        if key_to_check in request_can_dispatch_list:
            if len(request_can_dispatch_list[key_to_check]) > 0:
                return this_device_id

        # 2. check same level devices
        available_device_ids = []
        for device_id in same_level_device_ids:
            key_to_check = str(device_id)
            if key_to_check in request_can_dispatch_list:
                if len(request_can_dispatch_list[key_to_check]) > 0:
                    available_device_ids.append(device_id)

        if len(available_device_ids) > 0:
            return random.choice(available_device_ids)

        # 3. check child devices
        available_device_ids = []
        for device_id in child_device_ids:
            key_to_check = str(device_id)
            if key_to_check in request_can_dispatch_list:
                if len(request_can_dispatch_list[key_to_check]) > 0:
                    available_device_ids.append(device_id)

        if len(available_device_ids) > 0:
            return random.choice(available_device_ids)

        # 4. check parent devices
        available_device_ids = []
        for device_id in parent_device_ids:
            key_to_check = str(device_id)
            if key_to_check in request_can_dispatch_list:
                if len(request_can_dispatch_list[key_to_check]) > 0:
                    available_device_ids.append(device_id)

        if len(available_device_ids) > 0:
            return random.choice(available_device_ids)

        # 5. check parents' children
        available_device_ids = []
        for parent_device_id in parent_device_ids:
            for network_device in self.network_devices:
                if network_device["id"] == parent_device_id:
                    for device_id in network_device["child_device_ids"]:
                        key_to_check = str(device_id)
                        if key_to_check in request_can_dispatch_list:
                            if len(request_can_dispatch_list[key_to_check]) > 0:
                                available_device_ids.append(device_id)

        if len(available_device_ids) > 0:
            return random.choice(available_device_ids)


        # 6. check cloud
        available_device_ids = []
        for network_device in self.network_devices:
            if network_device["identity"] == "cloud":
                key_to_check = str(network_device["id"])
                if key_to_check in request_can_dispatch_list:
                    if len(request_can_dispatch_list[key_to_check]) > 0:
                        available_device_ids.append(device_id)

        if len(available_device_ids) > 0:
            return random.choice(available_device_ids)

        print("Error - Service Discovery Information Missing or there is no service instance available.")
        return -1

File: src/main/java/org/serviceProvider/capacities/RequestDispatchingSimple.java
package org.serviceProvider.capacities;

import org.cloudbus.cloudsim.Log;
import org.enduser.networkPacket.NetworkPacket;
import org.infrastructureProvider.entities.NetworkDevice;

import java.util.*;

public class RequestDispatchingSimple extends RequestDispatchingRule {

    public RequestDispatchingSimple(ArrayList<NetworkDevice> networkDevices) {
        super(networkDevices);
    }

    /* 1. check self; 2. check same level devices; 3. check child devices; 4. check parent devices */

    /* return deviceId */
    @Override
    public int findDeviceId(NetworkPacket request, ServiceDiscovery serviceDiscovery,
                            int deviceId, ArrayList<Integer> childDeviceIds, ArrayList<Integer> parentDeviceIds, ArrayList<Integer> sameLevelDeviceIds) {

        int requestServiceId = request.getDestinationServiceId();
        Log.printLine("find service " + requestServiceId);

        if (!serviceDiscovery.getServiceIdToInstanceList().containsKey(requestServiceId)) {
            System.out.println("Service Discovery Information Missing or there is no service instance");
            return -1;
        } else {
            // 1. check self
            if (serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).containsKey(deviceId)) {

                if (!serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).get(deviceId).isEmpty()) {
                    return deviceId;
                }
            }
            // 2. check same level devices
            ArrayList<Integer> devices = new ArrayList<>();
            for (Integer sameLevelDeviceId : sameLevelDeviceIds) {
                if (serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).containsKey(sameLevelDeviceId)) {

                    if (!serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).get(sameLevelDeviceId).isEmpty()) {
                        devices.add(sameLevelDeviceId);
                    }
                }
            }
            if (!devices.isEmpty()) {
                Random r = new Random();
                int index = r.nextInt(devices.size());
                Log.printLine("service has find in" + devices.get(index));
                return devices.get(index);
            }

            // 3. check child devices
            ArrayList<Integer> devices1 = new ArrayList<>();
            for (Integer childDeviceId : childDeviceIds) {
                if (serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).containsKey(childDeviceId)) {

                    if (!serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).get(childDeviceId).isEmpty()) {
                        devices1.add(childDeviceId);
                    }
                }
            }
            if (!devices1.isEmpty()) {
                Random r = new Random();
                int index = r.nextInt(devices1.size());
                return devices1.get(index);
            }

            // 4. check parent devices
            ArrayList<Integer> devices2 = new ArrayList<>();
            for (Integer parentDeviceId : parentDeviceIds) {
                if (serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).containsKey(parentDeviceId)) {

                    if (!serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).get(parentDeviceId).isEmpty()) {
                        devices2.add(parentDeviceId);
                    }
                }
            }
            if (!devices2.isEmpty()) {
                Random r = new Random();
                int index = r.nextInt(devices2.size());
                return devices2.get(index);
            }

            // 5. check parents' children
            Set<Integer> hashSet = new HashSet<>();
            for (Integer parentDeviceId : parentDeviceIds) {
                ArrayList<Integer> parentChildIds = findDeviceAccDeviceId(getNetworkDevices(), parentDeviceId).getChildDeviceIds();
                for (Integer parentChildId : parentChildIds) {
                    if (serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).containsKey(parentChildId)) {

                        if (!serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).get(parentChildId).isEmpty()) {
                            hashSet.add(parentChildId);
                        }
                    }
                }

            }
            ArrayList<Integer> devices3 = new ArrayList<>(hashSet);
            if (!devices3.isEmpty()) {
                Random r = new Random();
                int index = r.nextInt(devices3.size());
                return devices3.get(index);
            }

            // 6. check others
            int id = findCloud(getNetworkDevices());
//            ArrayList<Integer> devices4 = new ArrayList<>();
//
//            for (int i = 0 ; i < getNetworkDevices().size(); i ++){
//                if (getNetworkDevices().get(i).getId() != id){
//                    if (serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).containsKey(getNetworkDevices().get(i).getId())){
//
//                        if (serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).get(getNetworkDevices().get(i).getId()).size() > 0){
//                            devices4.add(getNetworkDevices().get(i).getId());
//                        }
//                    }
//                }
//            }
//            if (devices4.size() > 0){
//                Random r = new Random();
//                int index = r.nextInt(devices4.size());
//                return devices4.get(index);
//            }

            // 7. check cloud
            if (serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).containsKey(id)) {
                if (!serviceDiscovery.getServiceIdToInstanceList().get(requestServiceId).get(id).isEmpty()) {
                    return id;
                }
            }

        }
        System.out.println("Service Discovery Information Missing or there is no service instance");
        return -1;
    }

    public NetworkDevice findDeviceAccDeviceId(ArrayList<NetworkDevice> networkDevices, int deviceId) {
        for (NetworkDevice device : networkDevices) {
            if (device.getId() == deviceId) {
                return device;
            }
        }
        return null;
    }

    public int findCloud(ArrayList<NetworkDevice> networkDevices) {
        for (NetworkDevice device : networkDevices) {
            if (Objects.equals(device.getIdentify(), "cloud")) {
                return device.getId();
            }
        }
        return -1;
    }


}
